src/lib/comms/sol-mqtt.c

```diff
@@ -0,0 +1,579 @@
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <errno.h>
+#include <mosquitto.h>
+#include <stdbool.h>
+
+#define SOL_LOG_DOMAIN &_sol_mqtt_log_domain
+#include <sol-log-internal.h>
+#include <sol-mainloop.h>
+#include <sol-util.h>
+
+#include "sol-mqtt.h"
+
+SOL_LOG_INTERNAL_DECLARE(_sol_mqtt_log_domain, "mqtt");
+
+static int init_ref;
+#define CHECK_INIT(ret) \
+    do { \
+        if (unlikely(init_ref < 1)) { \
+            SOL_WRN("sol-mqtt used before initialization"); \
+            return ret; \
+        } \
+    } while (0)
+
+#define MQTT_CHECK_API(ptr, ...) \
+    do {                                        \
+        if (unlikely(ptr->api_version != \
+            SOL_MQTT_CONFIG_API_VERSION)) { \
+            SOL_WRN("Couldn't handle mqtt handler that has unsupported " \
+                "version '%u', expected version is '%u'", \
+                ptr->api_version, SOL_MQTT_CONFIG_API_VERSION); \
+            return __VA_ARGS__; \
+        } \
+    } while (0)
+
+struct sol_mqtt {
+    struct mosquitto *mosq;
+
+    struct sol_fd *socket_read;
+    struct sol_fd *socket_write;
+
+    struct sol_timeout *connect_timeout;
+    struct sol_timeout *disconnect_timeout;
+    struct sol_timeout *publish_timeout;
+    struct sol_timeout *subscribe_timeout;
+    struct sol_timeout *unsubscribe_timeout;
+    struct sol_timeout *message_timeout;
+
+    const struct sol_mqtt_config *config;
+
+    void *data;
+
+    int socket_fd;
+
+    int connection_status;
+
+    time_t keepalive;
+};
+
+struct sol_mqtt_message {
+    char *topic;
+    void *payload;
+    size_t payload_len;
+    int id;
+    sol_mqtt_qos qos;
+    bool retain;
+};
+
+static void
+sol_mqtt_init(void)
+{
+    SOL_LOG_INTERNAL_INIT_ONCE;
+    init_ref++;
+    mosquitto_lib_init();
+}
+
+static void
+sol_mqtt_shutdown(void)
+{
+    init_ref--;
+
+    if (init_ref == 0)
+        mosquitto_lib_cleanup();
+}
+
+static bool
+sol_mqtt_event_loop(void *data, int fd, unsigned int active_flags)
+{
+    struct sol_mqtt *mqtt = data;
+    int r;
+
+    r = mosquitto_loop_read(mqtt->mosq, 1);
+    r |= mosquitto_loop_write(mqtt->mosq, 1);
+    r |= mosquitto_loop_misc(mqtt->mosq);
+
+    if (r != 0) {
+        SOL_WRN("Unable to perform mqtt socket operation");
+        goto remove;
+    }
+
+    /* No more data to write, remove SOL_FD_FLAGS_OUT event watcher */
+    if (active_flags & SOL_FD_FLAGS_OUT && !mosquitto_want_write(mqtt->mosq))
+        goto remove;
+
+    return true;
+
+remove:
+    if (active_flags & SOL_FD_FLAGS_OUT)
+        mqtt->socket_write = NULL;
+    else
+        mqtt->socket_read = NULL;
+
```

Comment:
I mean, in the one for `read`:

```c
r = mosquito_loop_read(mqtt->mosq, 1);
r |= mosquitto_loop_misc(mqtt->mosq); /* NO IDEA WHAT IT DOES */
if (!r) {
   mqtt->socket_read = NULL;
   return false;
}
return true;
```

And for `write`:
```c
r = mosquitto_loop_write(mqtt->mosq, 1);
r |= mosquitto_loop_misc(mqtt->mosq); /* NO IDEA WHAT IT DOES */
if (!r || !mosquitto_want_write(mqtt->mosq)) {
   mqtt->socket_read = NULL;
   return false;
}
return true;
```

If this brings and ordering issue due read/write in mosquitto, then use a single `fd_add` and handler, flagging what we're doing:

```c
r = 0;
if (active_flags & SOL_FD_FLAGS_IN)
   r |= mosquitto_loop_read(mqtt->mosq, 1);
if (active_flags & SOL_FD_FLAGS_OUT)
  r |= mosquitto_loop_write(mqtt->mosq, 1);

r |= mosquitto_loop_misc(mqtt->mosq);

if (!r) {
   mqtt->fd_handler = NULL;
   return false;
}

wants_write = mosquitto_want_write(mqtt->mosq);
if (mqtt->watching_write && !wants_write) {
    mqtt->fd_handler = sol_fd_handler_add(mqtt->fd, SOL_FD_FLAGS_IN... );
    mqtt->watching_write = false;
    return false; /* delete the previous handler that also did write */
} else if (!mqtt->watching_write && wants_write) {
    mqtt->fd_handler = sol_fd_handler_add(mqtt->fd, SOL_FD_FLAGS_IN | SOL_FD_FLAGS_OUT... );
    mqtt->watching_write = true;
    return false; /* delete the previous handler that did not write */
}
```

Alternatively we can introduce a fd_set_flags() to allow changing the flags... should be more efficient.